File: util/quicksort.js
function quicksort(arr) {
	function swap(i, j) {
		const temp = arr[i];
		arr[i] = arr[j];
		arr[j] = temp;
	}

	function partition(start, end) {
		let pivot = arr[end];
		let pivot_index = start;

		for (let i = start; i < end; i++) {
			if (arr[i] < pivot) {
				swap(i, pivot_index);
				pivot_index++;
			}

			swap(pivot_index, end);
			return pivot_index;
		}
	}

	function quicksort_helper(start, end) {
		if (start >= end) return;
		const pivot_index = partition(start, end);
		quicksort_helper(start, pivot_index - 1);
		quicksort_helper(pivot_index + 1, end);
	}

	return quicksort_helper(0, arr.length - 1);
}

function expect(a, b) {
	if (a !== b) {
		console.log("Expected ", a, " but got ", b);
	}
}

const arr = [3, 2, 1, 4, 5];
quicksort(arr);

expect(arr, [1, 2, 3, 4, 5]);

File: util/quicksort.go
package util

func QuickSort(arr []int) {
	partition := func(i, j int) int {
		arr[i], arr[j] = arr[j], arr[i]
		pivot := arr[j]

		k := i
		for p := i; p < j; p++ {
			if arr[p] < pivot {
				arr[p], arr[k] = arr[k], arr[p]
				k++
			}
		}

		arr[k], arr[j] = arr[j], arr[k]
		return k
	}

	var sort func(int, int)
	sort = func(i, j int) {
		if i >= j {
			return
		}

		p := partition(i, j)
		sort(i, p-1)
		sort(p+1, j)
	}

	sort(0, len(arr)-1)
}
